lib/ansible/modules/cloud/azure/azure_vm_metadata_facts.py

```diff
@@ -0,0 +1,369 @@
+            sample:      ""
+
+        azure_vm_compute_version:
+            description: Azure VM version
+            type:        str
+            sample:      "18.04.201904020"
+
+        azure_vm_compute_vmid:
+            description: Azure VM id
+            type:        str
+            sample:      "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
+
+        azure_vm_compute_vmscalesetname:
+            description: Azure VM vm scaleset name
+            type:        str
+            sample:      ""
+
+        azure_vm_compute_vmsize:
+            description: Azure VM vm size
+            type:        str
+            sample:      "Standard_B1s"
+
+        azure_vm_compute_zone:
+            description: Azure VM zone
+            type:        str
+            sample:      ""
+
+        azure_vm_network_interface_ipv4_ipaddress_privateipaddress:
+            description: Azure VM network_interface_ipv4_ipaddress_privateipaddress
+            type:        str
+            sample:      "x.x.x.x"
+
+        azure_vm_network_interface_ipv4_ipaddress_publicipaddress:
+            description: Azure VM network_interface_ipv4_ipaddress_publicipaddress
+            type:        str
+            sample:      ""
+
+        azure_vm_network_interface_ipv4_subnet_address:
+            description: Azure VM network_interface_ipv4_subnet_address
+            type:        str
+            sample:      "x.x.x.0"
+
+        azure_vm_network_interface_ipv4_subnet_prefix:
+            description: Azure VM network_interface_ipv4_subnet_prefix
+            type:        str
+            sample:      "24"
+
+        azure_vm_network_interface_macaddress:
+            description: Azure VM network_interface_macaddress
+            type:        str
+            sample:      "XXXXXXXXXXXX"
+
+'''
+
+import json
+import re
+import socket
+import time
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.urls import fetch_url
+from ansible.module_utils.six.moves.urllib.parse import quote
+
+socket.setdefaulttimeout(5)
+
+
+class AzureVmMetadata(object):
+    #
+    # Metadata URL comes from:
+    # https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service
+    #
+    azure_vm_metadata_uri = 'http://169.254.169.254/metadata/instance?api-version=2019-02-01'
+
+    def __init__(self, module, azure_vm_metadata_uri=None):
+        self.module = module
+        self.uri_meta = azure_vm_metadata_uri or self.azure_vm_metadata_uri
+        self._data = {}
+        self._prefix = 'ansible_azure_vm%s'
+
+    #
+    # Query the metadata URL and return the results as text.
+    #
+    def _fetch(self, url):
+        encoded_url = quote(url, safe='%/:=&?~#+!$,;\'@()*[]')
+        response, info = fetch_url(self.module, encoded_url, method="GET", force=True, headers={'Metadata': 'true'})
+        if info['status'] not in (200, 404):
+            time.sleep(3)
```

Comment:
Is there any reason to add the retry logic?

Reply:
We can't assume that the azure metadata service will work on the first try.
The retry accommodates any small hiccups.
This is the same behavior as the AWS EC2 metadata module.